### PR TITLE
gluster interface for gd1 and gd2

### PR DESCRIPTION
**Pull request by vredara**
added interface to run specific cmds against gd1 and gd2
gd1.go has gd1 related functions, gd2 has glusterd2 related functions
glusterutils.go has interface which gd1 and gd2 has to implement

Signed-off-by: venkata edara <redara@redhat.com>

### CODE REVIEW

File: pkg/glusterutils/glusterutils.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
+package glusterutils
```
Comment (shtripat) [inline pkg/glusterutils/glusterutils.go]: I feel this is candidate for `glusterutils/exported.go`

**Comment by vredara** [comment]
I merged the vol info into respective interfaces. I also added MakeGluster func to create GD1 or GD2 based on config. all gd1 related can go to gd1.go and gd2 related funcs can be in gd2.go.

**Comment by vredara** [comment]
rebased it with master and did commit again.

File: pkg/glusterutils/exporterd.go [hunk 2]
```diff
@@ -13,25 +14,14 @@ var (
...
-
-	return peersGD1(config)
-}
-
 // IsLeader returns true or false based on whether the node is the leader of the cluster or not
-func IsLeader(config *Config) (bool, error) {
-	setDefaultConfig(config)
-	peerList, err := Peers(config)
+func (g GD1) IsLeader(config *Config) (bool, error) {
+	setDefaultConfig(g.config)
```
Comment (aravindavk) [inline pkg/glusterutils/exporterd.go]: `setDefaultConfig` can be moved to `MakeGluster` since it is required by all calls

File: pkg/glusterutils/exporterd.go [hunk 3]
```diff
@@ -13,25 +14,14 @@ var (
...
-	}
-
-	return peersGD1(config)
-}
-
 // IsLeader returns true or false based on whether the node is the leader of the cluster or not
-func IsLeader(config *Config) (bool, error) {
-	setDefaultConfig(config)
-	peerList, err := Peers(config)
+func (g GD1) IsLeader(config *Config) (bool, error) {
```
Comment (aravindavk) [inline pkg/glusterutils/exporterd.go]: - config parameter can be removed since already available as `g.config`
- change `(g GD1)` to `(g *GD1)`

File: pkg/glusterutils/exporterd.go [hunk 4]
```diff
@@ -13,25 +14,14 @@ var (
...
+	setDefaultConfig(g.config)
+	peerList, err := g.Peers()
 	if err != nil {
 		return false, err
 	}
-	peerID, err := LocalPeerID(config)
+	peerID, err := LocalPeerID(g.config)
```
Comment (aravindavk) [inline pkg/glusterutils/exporterd.go]: Also add `LocalPeerID` to interface as mentioned earlier. and use `peerID, err := g.LocalPeerID()`

File: pkg/glusterutils/exporterd.go [hunk 5]
```diff
@@ -22,6 +23,15 @@ func VolumeInfo(config *Config) ([]Volume, error) {
 
 	return gd1VolumeInfo(config)
 }
+*/
+
+func MakeGluster(config *Config) ginterface {
+	if config.GlusterMgmt == "" || config.GlusterMgmt == "glusterd" {
```
[the remark this answers is not in the document]
Comment (aravindavk) [reply pkg/glusterutils/exporterd.go]: `if config.GlusterMgmt == MgmtGlusterd`

File: pkg/glusterutils/exporterd.go [hunk 6]
```diff
@@ -64,15 +54,12 @@ func IsLeader(config *Config) (bool, error) {
...
 
-// VolumeInfo gets Volume info from Gd1/Gd2 based on config provided
-func VolumeInfo(config *Config) ([]Volume, error) {
-	setDefaultConfig(config)
-
-	if config.GlusterMgmt == MgmtGlusterd2 {
-		return gd2VolumeInfo(config)
+func MakeGluster(config *Config) ginterface {
+	if config.GlusterMgmt == "" || config.GlusterMgmt == "glusterd" {
+		return GD1{config: config}
```
Comment (aravindavk) [inline pkg/glusterutils/exporterd.go]: Also call `setDefaultConfig(config)`

```
setDefaultConfig(config)
if config.GlusterMgmt == MgmtGlusterd {
    return GD1{config: config}
}
return GD2{config: config}
```

File: pkg/glusterutils/exporterd.go [hunk 7]
```diff
@@ -56,3 +66,44 @@ func LocalPeerID(config *Config) (string, error) {
...
+func GetHealInfo(g ginterface, vol string) ([]HealEntry, error) {
+	return g.GetHealInfo(vol)
+}
+
+func GetVolumeInfo(g ginterface, vol string) ([]Volume, error) {
+	return g.GetVolumeInfo(vol)
+}
+
+// GetGlusterVersion gets the glusterfs version
+func GetGlusterVersion() (string, error) {
```
[the remark this answers is not in the document]
Comment (aravindavk) [reply pkg/glusterutils/exporterd.go]: this comment not addressed

File: pkg/glusterutils/exporterd.go [hunk 8]
```diff
@@ -114,3 +101,24 @@ func LocalPeerID(config *Config) (string, error) {
...
+	if err != nil {
+		return "", err
+	}
+	stdout := string(bytes[:])
+	fields := strings.Fields(stdout)
+	return fields[1], err
+}
+
+// ExecuteCmd enables to execute system cmds and returns stdout, err
+func ExecuteCmd(cmd string) ([]byte, error) {
```
Comment (aravindavk) [inline pkg/glusterutils/exporterd.go]: exposing this function not required. change name to `executeCmd`

File: pkg/glusterutils/types.go [hunk 9]
```diff
@@ -88,3 +88,28 @@ type Volume struct {
...
+	Name           string
+	Connected      string
+	NumHealEntries int
+}
+
+// ginterface should be implemented in GD1 and GD2 structs
+type ginterface interface {
+	Peers() ([]Peer, error)
+	HealInfo(vol string) ([]HealEntry, error)
+	VolumeInfo(vol string) ([]Volume, error)
```
Comment (aravindavk) [inline pkg/glusterutils/types.go]: add `IsLeader` and `LocalPeerID`

File: pkg/glusterutils/exporterd.go (hunk 3 above)
Comment (aravindavk) [reply pkg/glusterutils/exporterd.go]: We also need `func (g *GD2) IsLeader() (bool, error)`

```
func (g *GD1) IsLeader() (bool, error) {
	peerList, err := Peers(config)
	if err != nil {
		return false, err
	}
	peerID, err := LocalPeerID(config)
	if err != nil {
		return false, err
	}

	var maxPeerID string
	//This for loop iterates among all the peers and finds the peer with the maximum UUID (lexicographically)
	for i, pr := range peerList {
		if pr.Online {
			if peerList[i].ID > maxPeerID {
				maxPeerID = peerList[i].ID
			}
		}
	}
	//Checks and returns true if maximum peerID is equal to the local peerID
	if maxPeerID == peerID {
		return true, nil
	}
	return false, nil
}

func (g *GD2) IsLeader() (bool, error) {
	peerList, err := Peers(config)
	if err != nil {
		return false, err
	}
	peerID, err := LocalPeerID(config)
	if err != nil {
		return false, err
	}
	//The following lines checks and returns true if the local PeerID is equal to that of the first PeerID in the list
	for _, pr := range peerList {
		if pr.Online {
			if peerID == peerList[0].ID {
				return true, nil
			}
			return false, nil
		}
	}
	// This would imply none of the peers are online and return false
	return false, nil
}
```

File: pkg/glusterutils/healinfo_gd1.go [hunk 10]
```diff
@@ -0,0 +1,33 @@
+package glusterutils
+
+import (
+	"encoding/xml"
+	"fmt"
+	"strconv"
+)
+
+// GetHealInfo gets gluster vol heal in []HealEntries
+func (g GD1) HealInfo(vol string) ([]HealEntry, error) {
```
Comment (aravindavk) [inline pkg/glusterutils/healinfo_gd1.go]: change `(g GD1)` to `(g *GD1)`

File: pkg/glusterutils/healinfo_gd2.go [hunk 11]
```diff
@@ -0,0 +1,21 @@
+package glusterutils
+
+// GetHealInfo gets heal info from glusterd2 using rest api
+func (g GD2) HealInfo(vol string) ([]HealEntry, error) {
```
Comment (aravindavk) [inline pkg/glusterutils/healinfo_gd2.go]: change `(g GD2)` to `(g *GD2)`

File: pkg/glusterutils/peers_gd1.go [hunk 12]
```diff
@@ -19,10 +19,10 @@ type peersGlusterd1 struct {
 }
 
 // PeersGD1 returns the list of peers ( for GlusterD1 )
-func peersGD1(config *Config) ([]Peer, error) {
+func (g GD1) Peers() ([]Peer, error) {
```
Comment (aravindavk) [inline pkg/glusterutils/peers_gd1.go]: change `(g GD1)` to `(g *GD1)`

File: pkg/glusterutils/peers_gd2.go [hunk 13]
```diff
@@ -9,9 +9,9 @@ var (
 )
 
 // PeersGD2 returns the list of peers ( for GlusterD2 )
-func peersGD2(config *Config) ([]Peer, error) {
+func (g GD2) Peers() ([]Peer, error) {
```
Comment (aravindavk) [inline pkg/glusterutils/peers_gd2.go]: change `(g GD2)` to `(g *GD2)`

File: pkg/glusterutils/volinfo_gd1.go [hunk 14]
```diff
@@ -7,77 +7,9 @@ import (
...
-	}
-
-	if disperseCount > 0 {
-		return disperseCount
-	}
-	return 1
-}
-
-func gd1VolumeInfo(config *Config) ([]Volume, error) {
+func (g GD1) VolumeInfo(volname string) ([]Volume, error) {
```
Comment (aravindavk) [inline pkg/glusterutils/volinfo_gd1.go]: change `(g GD1)` to `(g *GD1)`

File: pkg/glusterutils/volinfo_gd2.go [hunk 15]
```diff
@@ -11,12 +11,12 @@ var subvolTypeValueToName = map[api.SubvolType]string{
 	api.SubvolDisperse:   "Disperse",
 }
 
-func gd2VolumeInfo(config *Config) ([]Volume, error) {
-	client, err := initRESTClient(config)
+func (g GD2) VolumeInfo(volname string) ([]Volume, error) {
```
Comment (aravindavk) [inline pkg/glusterutils/volinfo_gd2.go]: change `(g GD2)` to `(g *GD2)`

File: pkg/glusterutils/volinfo_gd1.go [hunk 16]
```diff
@@ -7,77 +7,9 @@ import (
...
 	// Run Gluster volume info --xml
-	out, err := exec.Command(config.GlusterCmd, "volume", "info", "--xml").Output()
+	out, err := exec.Command(g.config.GlusterCmd, "volume", "info", volname, "--xml").Output()
```
Comment (aravindavk) [inline pkg/glusterutils/volinfo_gd1.go]: empty string in argument may be problem try running `gluster volume info "" --xml`

File: pkg/glusterutils/healinfo_gd1.go [hunk 17]
```diff
@@ -0,0 +1,33 @@
+package glusterutils
```
Comment (shtripat) [inline pkg/glusterutils/healinfo_gd1.go]: I would suggest heal info to be done in another PR and not mix in this.
Lets take the interface changes in one PR first and then send another PR for heal info.

Comment (vredara) [reply pkg/glusterutils/healinfo_gd1.go]: ack

File: gluster-exporter/metric_brick.go [hunk 18]
```diff
@@ -129,14 +129,15 @@ func diskUsage(path string) (disk DiskStatus, err error) {
 }
 
 func brickUtilization() {
-	volumes, err := glusterutils.VolumeInfo(&glusterConfig)
+	gluster := glusterutils.MakeGluster(&glusterConfig)
```
Comment (aravindavk) [inline gluster-exporter/metric_brick.go]: This can be moved to `main.go`. Make `gluster` as global variable similar to `glusterConfig`

File: gluster-exporter/metric_ps.go [hunk 19]
```diff
@@ -71,6 +71,8 @@ var (
 		},
 		labels,
 	)
+
+	gluster = glusterutils.MakeGluster(&glusterConfig)
```
Comment (aravindavk) [inline gluster-exporter/metric_ps.go]: same as above

File: pkg/glusterutils/exporterd.go [hunk 20]
```diff
@@ -63,28 +40,73 @@ func IsLeader(config *Config) (bool, error) {
...
-	peeridFile := config.GlusterdWorkdir + "/glusterd.info"
-	if config.GlusterMgmt == MgmtGlusterd2 {
-		keywordID = "peer-id"
-		peeridFile = config.GlusterdWorkdir + "/uuid.toml"
+	peeridFile := g.config.GlusterdWorkdir + "/glusterd.info"
+	fileStream, err := os.Open(filepath.Clean(peeridFile))
+	if err != nil {
+		return "", err
 	}
+	defer func() {
```
Comment (aravindavk) [inline pkg/glusterutils/exporterd.go]: below code is duplicated. create one more internal function and call from both `GD1.LocalPeerID` and `GD2.LocalPeerID`

File: pkg/glusterutils/exporterd.go [hunk 21]
```diff
@@ -114,3 +136,24 @@ func LocalPeerID(config *Config) (string, error) {
...
+	bytes, err := executeCmd(cmd)
+	if err != nil {
+		return "", err
+	}
+	stdout := string(bytes[:])
+	fields := strings.Fields(stdout)
+	return fields[1], err
+}
+
+// ExecuteCmd enables to execute system cmds and returns stdout, err
```
Comment (aravindavk) [inline pkg/glusterutils/exporterd.go]: nit: code comment reflects old name

File: pkg/glusterutils/volinfo_gd1.go [hunk 22]
```diff
@@ -7,77 +7,9 @@ import (
...
-	}
-
-	if disperseCount > 0 {
-		return disperseCount
-	}
-	return 1
-}
-
-func gd1VolumeInfo(config *Config) ([]Volume, error) {
+func (g GD1) VolumeInfo() ([]Volume, error) {
```
Comment (aravindavk) [inline pkg/glusterutils/volinfo_gd1.go]: not changed to `(g *GD1)`

File: pkg/glusterutils/volinfo_gd2.go [hunk 23]
```diff
@@ -11,8 +11,8 @@ var subvolTypeValueToName = map[api.SubvolType]string{
 	api.SubvolDisperse:   "Disperse",
 }
 
-func gd2VolumeInfo(config *Config) ([]Volume, error) {
-	client, err := initRESTClient(config)
+func (g GD2) VolumeInfo() ([]Volume, error) {
```
Comment (aravindavk) [inline pkg/glusterutils/volinfo_gd2.go]: same here, not changed to `(g *GD2)`

File: gluster-exporter/metric_brick.go (hunk 18 above)
Comment (vredara) [reply gluster-exporter/metric_brick.go]: the actual config check is done in main() function in main.go. if i make gluster = MakeGluster(&config) before we parse the global conf, doesnt it always return gd1 instance?.

File: pkg/glusterutils/exporterd.go (hunk 20 above)
Comment (vredara) [reply pkg/glusterutils/exporterd.go]: ack

File: gluster-exporter/metric_brick.go (hunk 18 above)
Comment (aravindavk) [reply gluster-exporter/metric_brick.go]: it should be after parsing the config and before starting the metrics collectors

**Comment by aravindavk** [comment]
also please squash all commits into one.

File: gluster-exporter/metric_brick.go (hunk 18 above)
Comment (vredara) [reply gluster-exporter/metric_brick.go]: I just declared gluster var as global and assigned it in main().

**Comment by vredara** [comment]
Incorporated all comments.

File: pkg/glusterutils/exporterd.go [hunk 24]
```diff
@@ -114,3 +98,101 @@ func LocalPeerID(config *Config) (string, error) {
...
+
+// LocalPeerID returns local peer ID of glusterd
+func (g *GD1) LocalPeerID() (string, error) {
+	keywordID := "UUID"
+	peeridFile := g.config.GlusterdWorkdir + "/glusterd.info"
+	fileStream, err := os.Open(filepath.Clean(peeridFile))
+	if err != nil {
+		return "", err
+	}
+	peerid, err := readPeerID(fileStream, keywordID)
```
Comment (aravindavk) [inline pkg/glusterutils/exporterd.go]: directly return `return readPeerID(fileStream, keywordID)`

File: pkg/glusterutils/exporterd.go [hunk 25]
```diff
@@ -114,3 +98,101 @@ func LocalPeerID(config *Config) (string, error) {
...
+	return peerid, err
+	/*
```
Comment (aravindavk) [inline pkg/glusterutils/exporterd.go]: please remove commented code

File: pkg/glusterutils/exporterd.go [hunk 26]
```diff
@@ -114,3 +98,101 @@ func LocalPeerID(config *Config) (string, error) {
...
+
+// LocalPeerID returns local peer ID of glusterd2
+func (g *GD2) LocalPeerID() (string, error) {
+	keywordID := "peer-id"
+	peeridFile := g.config.GlusterdWorkdir + "/uuid.toml"
+	fileStream, err := os.Open(filepath.Clean(peeridFile))
+	if err != nil {
+		return "", err
+	}
+	peerid, err := readPeerID(fileStream, keywordID)
```
Comment (aravindavk) [inline pkg/glusterutils/exporterd.go]: `return readPeerID(fileStream, keywordID)`

File: pkg/glusterutils/exporterd.go [hunk 27]
```diff
@@ -114,3 +98,101 @@ func LocalPeerID(config *Config) (string, error) {
...
+	return peerid, err
+
+	/*
```
Comment (aravindavk) [inline pkg/glusterutils/exporterd.go]: remove commented code

File: gluster-exporter/main.go [hunk 28]
```diff
@@ -28,6 +28,7 @@ var (
 	config                        = flag.String("config", defaultConfFile, "Config file path")
 	defaultInterval time.Duration = 5
 	glusterConfig   glusterutils.Config
+	gluster         glusterutils.ginterface
```
Comment (aravindavk) [inline gluster-exporter/main.go]: Unexported name used here

**Comment by vredara** [comment]
I could not see metric_volume_counts.go in my git env. I rebased the branch and fixed the build issues. turned ginterface to GInterface , now the name is exposed and can be used in main.go

**Comment by aravindavk** [comment]
Please check the build failure.

**Comment by vredara** [comment]
> Please check the build failure.

metric_volume_counts.go is not visible to me, so I  am not able to change code. I fetched the latest master , rebased but still metric_volume_counts.go is not there. will work on this.

**Comment by aravindavk** [comment]
```
git checkout master
git fetch upstream
git merge upstream/master
git checkout gluster-utils
git rebase -i master
git push origin gluster-utils -f
```

**Comment by aravindavk** [comment]
@vredara thanks for the PR
